File: solutions/cs_solutions/p20.cs
public class p20
{
     // Valid Parentheses - Easy

    // Given a string containing just the characters '(', ')', '{', '}', 
    //  '[' and ']', determine if the input string is valid.
    // An input string is valid if:
    //   Open brackets must be closed by the same type of brackets.
    //   Open brackets must be closed in the correct order.
    // Note that an empty string is also considered valid.
    public class Solution
    {
        public bool IsValid(string s)
        {
            if (s.Length == 0) return true;

            var closing = new Stack<char>();

            foreach (char c in s)
            {
                if (c == '[') closing.Push(']');
                else if (c == '(') closing.Push(')');
                else if (c == '{') closing.Push('}');
                else if (closing.Count == 0 || closing.Pop() != c)
                    return false;
            }
            return closing.Count == 0;
        }
    }
}

File: solutions/cpp_solutions/p20.cpp
#include <string>
#include <vector>
using namespace std;

// Valid Parentheses - Easy

// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.

class Solution {
public:
    bool isValid(string s) {
        if (s.size() == 0) return true;

        vector<char> closing;
        for (char c : s){
            if (c == '[')
                closing.push_back(']');
            else if (c == '{')
                closing.push_back('}');
            else if (c == '(')
                closing.push_back(')');
            else {
                if (closing.size() == 0)
                    return false;
                if (closing.back() != c)
                    return false;
                closing.pop_back();
            }
        }

        return closing.size() == 0;
    }
};
